compiler/src/lnntoamm/Types.ts

```diff
@@ -123,127 +193,279 @@ export default abstract class Type implements Equalable {
       );
-    } else if (ty instanceof HasField) {
-      throw new Error(`type ${this.name} does not have field ${ty.name}`);
-    } else if (!this.compatibleWithConstraint(ty, scope)) {
-      throw new Error(
-        `type ${this.name} could not be constrained to ${ty.name}`,
+    }
+    if (that instanceof Opaque) {
+      if (
+        Object.values(this.generics).some((g) => g === null) ||
+        Object.values(that.generics).some((g) => g === null)
+      ) {
+        // if any values are null values, that means that we just have to check
+        // for constraint compatibilities which was already checked so we're good
+        return;
+      }
+      const thisGens = Object.keys(this.generics);
+      const thatGens = Object.keys(that.generics);
+      thisGens.forEach((genName, ii) =>
+        this.generics[genName].constrain(that.generics[thatGens[ii]], scope),
       );
+    } else if (that instanceof Interface || that instanceof OneOf) {
+      that.constrain(this, scope);
+    } else {
+      console.log(this);
+      console.log(that);
+      throw 'uh';
     }
   }
 
-  eq(that: Equalable): boolean {
-    if (that instanceof Builtin) {
-      return this === that;
-    } else if (that instanceof Interface) {
-      if (that instanceof Generated && that.delegate !== null) {
-        return this.eq(that.delegate);
+  contains(that: Type): boolean {
+    return this.eq(that);
+  }
+
+  dup(): Type | null {
+    const genKeys = Object.keys(this.generics);
+    if (genKeys.length === 0) {
+      return null;
+    }
+    const duped = new Opaque(this.name, genKeys);
+    let isNothingNew = true;
+    genKeys.forEach((genName) => {
+      const thisGen = this.generics[genName];
+      let tyVal: Type;
+      if (thisGen === null) {
+        tyVal = Type.generate();
+      } else {
+        const duped = thisGen.dup();
+        if (duped === null) {
+          tyVal = thisGen;
+        } else {
+          tyVal = duped;
+          isNothingNew = false;
+        }
       }
-      return that.tempDelegate !== null && this.eq(that.tempDelegate);
-    } else {
+      duped.generics[genName] = tyVal;
+    });
+    return duped;
+  }
+
+  eq(that: Equalable): boolean {
+    if (!(that instanceof Opaque) || this.name !== that.name) {
       return false;
     }
+    const thisGens = Object.values(this.generics);
+    const thatGens = Object.values(that.generics);
+    return (
+      thisGens.length === thatGens.length &&
+      thisGens.every((thisGen, ii) => {
+        const thatGen = thatGens[ii];
+        if (thisGen === null || thatGen === null) {
+          return thisGen === thatGen;
+        } else {
+          return thisGen.eq(thatGen);
+        }
+      })
+    );
   }
 
-  fieldIndices(): FieldIndices {
-    return TODO("determine if it's even worth keeping the Builtin class");
+  fnselectOptions(): Type[] {
+    const genOptions = Object.values(this.generics).map((g) =>
+      g === null ? [g] : g.fnselectOptions(),
+    );
```

Comment:
```suggestion
    const genOptions = Object.values(this.generics).map((g) => g?.fnselectOptions() ?? [g]);
```

Reply:
> `g?.fnselectOptions()`

does this work? I can't find any information on the `?` operator when it's not used with `:`

Reply:
It's not the `?` operator, it's the [`?.` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)